File: calendar-shared/src/main/java/uk/co/unclealex/hammers/calendar/shared/model/Competition.java
/**
 * Copyright 2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 *
 */
package uk.co.unclealex.hammers.calendar.shared.model;

import java.util.Comparator;
import java.util.Map;
import java.util.SortedMap;
import java.util.TreeMap;


/**
 * The different competitions that West Ham have taken part in.
 * 
 * @author alex
 * 
 */
public enum Competition {
	/**
	 * The FA Premiership.
	 */
	PREM("Premiership", true, "PREM"),

	/**
	 * The League Cup.
	 */
	LGCP("League Cup", false, "LGCP"),

	/**
	 * The FA Cup.
	 */
	FACP("FA Cup", false, "FACP"),

	/**
	 * The Championship.
	 */
	FLC("Championship", true, "FLC", "FLD1"),

	/**
	 * The Championship play-offs.
	 */
	FLCPO("Play-Offs", false, "FLD1 P/O", "FLC P/O");

	/**
	 * Map all competitions by their token.
	 */
	private static SortedMap<String, Competition> COMPETITIONS_BY_TOKEN;
	static {
		Comparator<String> longestFirstComparator = new Comparator<String>() {
			@Override
			public int compare(String s1, String s2) {
				int cmp = s2.length() - s1.length();
				return cmp == 0 ? s1.compareTo(s2) : cmp;
			}
		};
		COMPETITIONS_BY_TOKEN = new TreeMap<String, Competition>(longestFirstComparator);
		for (Competition competition : Competition.values()) {
			for (String token : competition.getTokens()) {
				COMPETITIONS_BY_TOKEN.put(token, competition);
			}
		}
	}

	/**
	 * Get a competition by its token.
	 * @param token The token to look for.
	 * @return The competition for the token.
	 * @throws IllegalArgumentException Thrown if no such competition exists.
	 */
	public static Competition findByToken(String token) throws IllegalArgumentException {
		for (Map.Entry<String, Competition> entry : COMPETITIONS_BY_TOKEN.entrySet()) {
			if (token.startsWith(entry.getKey())) {
				return entry.getValue();
			}
		}
		throw new IllegalArgumentException(token + " is not a valid competition token.");
	}

	/**
	 * An array of tokens that can identify competition on the West Ham website.
	 */
	private String[] tokens;
	
	/**
	 * True if competition is a league competition, false otherwise.
	 */
	private boolean league;
	
	/**
	 * The name of the competition.
	 */
	private String name;

	/**
	 * Instantiates a new competition.
	 * 
	 * @param name
	 *          the name
	 * @param league
	 *          the league
	 * @param tokens
	 *          the tokens
	 */
	private Competition(String name, boolean league, String... tokens) {
		this.name = name;
		this.league = league;
		this.tokens = tokens;
	}

	/**
	 * Gets the name of the competition.
	 * 
	 * @return the name of the competition
	 */
	public String getName() {
		return name;
	}

	/**
	 * Gets the an array of tokens that can identify competition on the West
	 * Ham website.
	 * 
	 * @return the an array of tokens that can identify competition on the
	 *         West Ham website
	 */
	public String[] getTokens() {
		return tokens;
	}

	/**
	 * Checks if is true if competition is a league competition, false
	 * otherwise.
	 * 
	 * @return the true if competition is a league competition, false
	 *         otherwise
	 */
	public boolean isLeague() {
		return league;
	}
}

File: app/uk/co/unclealex/hammers/calendar/model/Competition.scala
/**
 * Copyright 2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation  extends Competition(0, ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0  extends Competition(0, the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package uk.co.unclealex.hammers.calendar.model

import uk.co.unclealex.hammers.calendar.logging.{RemoteLogging, RemoteStream}

/**
 * The different competitions that West Ham have taken part in.
 *
 * @author alex
 *
 */
sealed trait Competition extends Competition.Value {
  /**
   * The English name of this competition.
   */
  val name: String
  /**
   * The tokens that are used within the West Ham website to identify this competition.
   */
  val tokens: List[String]

  /**
   * A flag that determines whether this competition is a primary English league competition or not.
   */
  def isLeague: Boolean

}

/**
 * League based competitions.
 */
sealed trait LeagueCompetition extends Competition {
  override def isLeague = true
}

/**
 * Cup based competitions.
 */
sealed trait CupCompetition extends Competition {

  override def isLeague = false
}

object Competition extends PersistableEnumeration[Competition] with RemoteLogging {
  /**
   * The FA Premiership.
   */
  case object PREM extends LeagueCompetition {
    val persistableToken = "PREM"
    val name = "Premiership"
    val tokens = List("Barclays Premier League")
  }
  PREM

  /**
   * The League Cup.
   */
  case object LGCP extends CupCompetition {
    val persistableToken = "LGCP"
    val name = "League Cup"
    val tokens = List("English Capital One Cup")
  }
  LGCP

  /**
   * The FA Cup.
   */
  case object FACP extends CupCompetition {
    val persistableToken = "FACP"
    val name = "FA Cup"
    val tokens = List("English FA Cup")
  }
  FACP

  /**
   * The Championship.
   */
  case object FLC extends LeagueCompetition {
    val persistableToken = "FLC"
    val name = "Championship"
    val tokens = List("FLC", "FLD1")
  }
  FLC

  /**
   * The Championship play-offs.
   */
  case object FLCPO extends CupCompetition {
    val persistableToken = "FLCPO"
    val name = "Play-Offs"
    val tokens = List("FLD1 P/O", "FLC P/O")
  }
  FLCPO

  def apply(token: String)(implicit remoteStream: RemoteStream): Option[Competition] = {
    logOnEmpty(values.find(_.tokens contains token), s"$token is not a valid competition token.")
  }

}
